testnet/stacks-node/src/nakamoto_node/miner.rs

```diff
@@ -354,10 +352,21 @@ impl BlockMinerThread {
 
         let burn_election_height = self.burn_election_block.block_height;
 
+        let reward_cycle = self
+            .burnchain
+            .block_height_to_reward_cycle(burn_election_height)
+            .expect("FATAL: no reward cycle for sortition");
+
+        #[cfg(test)]
+        {
+            info!(
+                "---- Fetching reward info at height {} for cycle {} ----",
+                burn_election_height, reward_cycle
+            );
+        }
+
         let reward_info = match load_nakamoto_reward_set(
-            self.burnchain
-                .pox_reward_cycle(burn_election_height)
-                .expect("FATAL: no reward cycle for sortition"),
+            reward_cycle,
```

Comment:
Previously, using `pox_reward_cycle` would incorrectly fetch cycle 11's reward set at burn height 240, when it should be fetching cycle 12.

Even after updating this, I'm getting odd logs around the loaded reward set. Even when cycle 12 is loaded, the `load_nakamoto_reward_set` function loads cycle 11's reward set, even when 12 is passed as `reward_cycle`. I've confirmed that other endpoints, like `/v2/reward_set` return the correct new reward set after block 136, and the signers are behaving "as expected". The test ends up failing because the miner is unable to verify signer signatures after 240, because it's looking at the wrong reward set.

Relevant logs below. Notice that it says "reward set from written state ... stacks_block_height: 13, burn_header_height: 16", which is for cycle 11. The other logs (during cycle 11) have the "fetching preprocessed reward set" log. I'm still digging into how and why reward sets are read and written, but I could use some help here. You can also see that the fetch reward set has 5 signers (which cycle 11 has), instead of 4.

```
INFO [1722554117.559187] [testnet/stacks-node/src/nakamoto_node/miner.rs:362] [miner.7773e9821b13395ece1ac702a07612f4509faa522ed84beb043f7b25641c59f3] ---- Fetching reward info at height 240 for cycle 12 ----
INFO [1722554117.561183] [stackslib/src/chainstate/nakamoto/coordinator/mod.rs:230] [miner.7773e9821b13395ece1ac702a07612f4509faa522ed84beb043f7b25641c59f3] PoX reward set loaded from written block state, reward_set_block_id: e9892227ca874fdb4913ff1253ac1ff931e044eaef04b98f0da352b7b531e30d, burn_block_hash: 590ebe1b91fb0e55c34f22781490af6d63a580a33e8d99bae3e0cfeac97527d7, stacks_block_height: 13, burn_header_height: 216
INFO [1722554117.561194] [testnet/stacks-node/src/nakamoto_node/miner.rs:398] [miner.7773e9821b13395ece1ac702a07612f4509faa522ed84beb043f7b25641c59f3] ---- New reward set has 5 signers ----
```

Reply:
Oh that def sounds like a bug. Glad a test actually caught potentially something XD.

I was seeing weird off by one issues at some point in mock_sign_epoch_25. I thought it was a bug in my logic, but this makes me think it probably was a bug I managed to work around.

I load  
```rust
let mut current_burn_block_height = signer_test
        .running_nodes
        .btc_regtest_controller
        .get_headers_height();
```

And THEN I advance the burn block height. But I do NOT update the current_burn_block_height because then I got an off by one when comparing against what the node state was. I am wondering if this is what is happening internally in your reward set calls. i.e. the burn block height is lets say 240, but internally it fetches burn block height 239. I am not 100% sure though, but this was the very first thing that came to mind for me.

Reply:
Yep -- this change in `miner.rs` is correct (`pox_reward_cycle` is a dangerous function, it should probably be removed though I think this requires more discussion).

These changes get the integration test passing for me locally:

```diff
diff --git a/stackslib/src/chainstate/nakamoto/coordinator/mod.rs b/stackslib/src/chainstate/nakamoto/coordinator/mod.rs
index 17cfed5cd..15cc7f085 100644
--- a/stackslib/src/chainstate/nakamoto/coordinator/mod.rs
+++ b/stackslib/src/chainstate/nakamoto/coordinator/mod.rs
@@ -546,7 +546,7 @@ pub fn load_nakamoto_reward_set<U: RewardSetProvider>(
            "burnchain_height" => %anchor_block_sn.block_height);
 
     let reward_set = provider.get_reward_set_nakamoto(
-        prepare_end_height.saturating_sub(1),
+        prepare_end_height,
         chain_state,
         burnchain,
         sort_db,
diff --git a/testnet/stacks-node/src/tests/signer/mod.rs b/testnet/stacks-node/src/tests/signer/mod.rs
index 35db96c84..3948b0dc8 100644
--- a/testnet/stacks-node/src/tests/signer/mod.rs
+++ b/testnet/stacks-node/src/tests/signer/mod.rs
@@ -64,7 +64,7 @@ use crate::tests::nakamoto_integrations::{
     naka_neon_integration_conf, next_block_and_mine_commit, POX_4_DEFAULT_STACKER_BALANCE,
 };
 use crate::tests::neon_integrations::{
-    next_block_and_wait, run_until_burnchain_height, test_observer, wait_for_runloop,
+    get_chain_info, next_block_and_wait, run_until_burnchain_height, test_observer, wait_for_runloop
 };
 use crate::tests::to_addr;
 use crate::{BitcoinRegtestController, BurnchainController};
@@ -477,11 +477,14 @@ impl<S: Signer<T> + Send + 'static, T: SignerEventTrait + 'static> SignerTest<Sp
             .running_nodes
             .btc_regtest_controller
             .get_headers_height();
-        self.running_nodes
+        let block_height = get_chain_info(&self.running_nodes.conf).burn_block_height;
+        let rc = self.running_nodes
             .btc_regtest_controller
             .get_burnchain()
             .block_height_to_reward_cycle(block_height)
-            .unwrap()
+            .unwrap();
+        info!("Get current reward cycle: block_height = {block_height}, rc = {rc}");
+        rc
     }
 
     fn get_signer_index(&self, reward_cycle: u64) -> SignerSlotID {
diff --git a/testnet/stacks-node/src/tests/signer/v0.rs b/testnet/stacks-node/src/tests/signer/v0.rs
index 9a6d362e5..f08f0f87e 100644
--- a/testnet/stacks-node/src/tests/signer/v0.rs
+++ b/testnet/stacks-node/src/tests/signer/v0.rs
@@ -2136,8 +2136,8 @@ fn signer_set_rollover() {
     let current_signers = signer_test.get_reward_set_signers(new_reward_cycle);
     assert_eq!(current_signers.len(), new_num_signers as usize);
     for signer in current_signers.iter() {
-        assert!(signer_test_public_keys.contains(&signer.signing_key.to_vec()));
-        assert!(!new_signer_public_keys.contains(&signer.signing_key.to_vec()));
+        assert!(!signer_test_public_keys.contains(&signer.signing_key.to_vec()));
+        assert!(new_signer_public_keys.contains(&signer.signing_key.to_vec()));
     }
 
     info!("---- Mining a block to verify new signer set -----");
```

3 changes really:

1. For me locally, the test was failing before cycle 12: `btc_regtest_controller.get_headers_height()` was returning 1 block higher than the node had processed.
2. I think the assertions I changed may have just been unreached copy-pasta?
3. The actual implementation change: loading the reward set makes use of a burnchain height parameter which is byzantine at best, and has led to a bunch of back and forth about off-by-one errors, etc. The line I changed above used to be as it is in patch, but was changed because of an off-by-one error, but looking at the code now, the subtraction just seems incorrect here.

This all leaves me very uncertain about the current way nakamoto reward sets are being calculated (or rather, it reminded me of how uncertain I felt about it previously). I _think_ that the safest approach here would be for these nakamoto reward set functions to all take the reward cycle as a parameter, but there still needs to be some thought in the codebase about how to compute the reward cycle number at each call site.

Reply:
Thanks @kantai , I've just pushed that diff! Passing for me now as well.

Reply:
Ah this is the error I was seeing where the btc_regtest_controller.get_headers_height() was returning 1 higher than the node had processed. Should I be changing this in my mock_signing test? I had worked around it by calling it before processing a burn block but am wondering if this is a bandaid that covers an underlying problem.